src/qmapshack/poi/CPoiPOI.cpp

```diff
@@ -56,10 +65,139 @@ void CPoiPOI::draw(IDrawContext::buffer_t& buf)
+    mutex.lock(); //Does this have to be so broad?
+    for(const QString& category : categoryActivated.keys(true))
+    {
+        qreal s = 22. / 32; //For now reduces the poi icons to the same size as wpt icons. Later something more intelligent may be calculated
+        QPixmap icon("://icons/poi/png/" + category + ".n.32.png");
+        icon = icon.scaled(icon.size() * s, Qt::KeepAspectRatio, Qt::SmoothTransformation);
+        for(int minLonM10 = qFloor(xMin * RAD_TO_DEG * 10); minLonM10 < qCeil(xMax * RAD_TO_DEG * 10); minLonM10++)
+        {
+            for(int minLatM10 = qFloor(yMin * RAD_TO_DEG * 10); minLatM10 < qCeil(yMax * RAD_TO_DEG * 10); minLatM10++)
+            {
+                if(!loadedPOIs.contains(category) ||
+                   !loadedPOIs[category].contains(minLonM10) ||
+                   !loadedPOIs[category][minLonM10].contains(minLatM10))
+                {
+                    mutex.unlock();
+                    loadPOIsFromFile(category, minLonM10, minLatM10);
+                    mutex.lock();
+                }
+
+                for(const poi_t& poiToDraw : loadedPOIs[category][minLonM10][minLatM10])
+                {
+                    QPointF pt = poiToDraw.coordinates;
+                    poi->convertRad2Px(pt);
+
+                    QRectF r1 = icon.rect();
+                    r1.moveCenter(pt);
+
+                    p.drawRect(r1.adjusted(-1, -1, 1, 1));
+                    p.drawPixmap(r1.topLeft(), icon);
+                }
+            }
+        }
+    }
+    mutex.unlock();
+}
+
+void CPoiPOI::addTreeWidgetItems(QTreeWidget* widget)
+{
+    QMap<QString, QTreeWidgetItem*> parentMap;
+
+    for(const QString& category : tagMap.keys())
+    {
+        QStringList parentChild = category.split("_");
+        QString parent = parentChild[0];
+        QString child = parentChild.mid(1).join("_");
+        if(!parentMap.contains(parent))
+        {
+            auto parentItem = new QTreeWidgetItem(widget);
+            parentItem->setData(CPoiPropSetup::eTreeColumnCategory, Qt::DisplayRole, parent);
+            parentItem->setCheckState(CPoiPropSetup::eTreeColumnCheckbox, Qt::Unchecked);
+            parentItem->setFlags(parentItem->flags() | Qt::ItemIsAutoTristate);
+            parentMap[parent] = parentItem;
+        }
+        QTreeWidgetItem* item = new QTreeWidgetItem(parentMap[parent]);
+        item->setData(CPoiPropSetup::eTreeColumnParentCategory, Qt::DisplayRole, parent);
+        item->setData(CPoiPropSetup::eTreeColumnCategory, Qt::DisplayRole, child);
+        item->setCheckState(CPoiPropSetup::eTreeColumnCheckbox, categoryActivated[category] ? Qt::Checked : Qt::Unchecked);
+    }
+}
+
+void CPoiPOI::checkedStateChanged(QTreeWidgetItem * item, int columnChanged)
+{
+    if(columnChanged == CPoiPropSetup::eTreeColumnCheckbox)  //shouldn't happen, but still
+    {
+        QString parent = item->data(CPoiPropSetup::eTreeColumnParentCategory, Qt::DisplayRole).toString();
```

Comment:
QTreeWidgetItem::data() / setData() should be avoided as it ends up quickly in cryptic code. Defining your own item class that inherits QTreeWidgetItem is a much cleaner approach